pallets/stateful-storage/src/lib.rs

```diff
@@ -0,0 +1,797 @@
+		StalePageState,
+
+		/// Invalid Signature for payload
+		InvalidSignature,
+
+		/// The submitted proof has expired; the current block is less the expiration block
+		ProofHasExpired,
+
+		/// The submitted proof expiration block is too far in the future
+		ProofNotYetValid,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// An event for when an itemized storage is updated
+		ItemizedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an itemized storage is deleted
+		ItemizedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is updated
+		PaginatedPageUpdated {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before update
+			prev_content_hash: PageHash,
+			/// current content hash after update
+			curr_content_hash: PageHash,
+		},
+
+		/// An event for when an paginated storage is deleted
+		PaginatedPageDeleted {
+			/// message source id of storage owner
+			msa_id: MessageSourceId,
+			/// schema related to the storage
+			schema_id: SchemaId,
+			/// id of updated page
+			page_id: PageId,
+			/// previous content hash before removal
+			prev_content_hash: PageHash,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Applies the Add or Delete Actions on the requested Itemized page
+		///
+		/// # Events
+		/// * [`Event::ItemizedPageUpdated`]
+		/// * [`Event::ItemizedPageDeleted`]
+		///
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::apply_item_actions( actions.len() as u32 ,
+			actions.iter().fold(0, |acc, a| acc + match a {
+				ItemAction::Add { data } => data.len() as u32,
+				_ => 0,
+			}),
+		    0
+		))]
+		pub fn apply_item_actions(
+			origin: OriginFor<T>,
+			#[pallet::compact] state_owner_msa_id: MessageSourceId,
+			#[pallet::compact] schema_id: SchemaId,
+			#[pallet::compact] target_hash: PageHash,
+			actions: BoundedVec<ItemAction, T::MaxItemizedActionsCount>,
+		) -> DispatchResult {
+			let provider_key = ensure_signed(origin)?;
+			let is_pruning = actions.iter().any(|a| matches!(a, ItemAction::Delete { .. }));
+			Self::check_actions(&actions)?;
+			Self::check_schema(schema_id, PayloadLocation::Itemized, false, is_pruning)?;
+			Self::check_grants(provider_key, state_owner_msa_id, schema_id)?;
```

Comment:
I think we should probably move this above check schema since you need an MSA to do anything. I might even move it up before check_actions but those do not require a db read.

Reply:
The function name might be misleading or doing too much. Because it also checks to see if the key is associated with a valid MSA id. This is the root of being able to do anything.

```
let provider_msa_id = T::MsaInfoProvider::ensure_valid_msa_key(&provider_key)
			.map_err(|_| Error::<T>::InvalidMessageSourceAccount)?;
```